walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,220 @@
+    List<TimeRange> optionalAttendeesBusyTimeRanges = getAllBusyTimeRanges(events, optionalAttendees);
+
+    sortBusyTimeRangesByStart(mandatoryAttendeesBusyTimeRanges); 
+    sortBusyTimeRangesByStart(optionalAttendeesBusyTimeRanges);
+
+    List<TimeRange> mandatoryAttendeesEmptyRanges = getEmptyTimeRanges(mandatoryAttendeesBusyTimeRanges, meetingDuration); 
+    List<TimeRange> optionalAttendeesEmptyRanges = getEmptyTimeRanges(optionalAttendeesBusyTimeRanges, meetingDuration);
+
+    if (mandatoryAttendees.isEmpty()) {
+      return optionalAttendeesEmptyRanges; 
+    } else if (optionalAttendees.isEmpty()) {
+      return mandatoryAttendeesEmptyRanges; 
+    } else {
+      List<TimeRange> allEmptyTimeRanges = findCommonEmptyTimeRanges(
+          mandatoryAttendeesEmptyRanges, optionalAttendeesEmptyRanges, meetingDuration);
+      return allEmptyTimeRanges;
+    }
+  }
+
+  /** 
+   * Constructs all the time ranges that any of the mandatory attendees is busy.
+   * These time ranges cannot be part of the potential meeting time. 
+   */ 
+  private List<TimeRange> getAllBusyTimeRanges(Collection<Event> events, Collection<String> mandatoryAttendees) {
+    List<TimeRange> busyTimes = new ArrayList<TimeRange>();
+
+    events.forEach((event) -> {
+      Collection<String> eventAttendees = event.getAttendees();
+
+      for (String attendee : eventAttendees) {
+        // If the current event's attendee list contains a mandatory attendee, 
+        // adds this event's time range to the set of all busy times, 
+        // after checking for overlapping time ranges. 
+        if (mandatoryAttendees.contains(attendee)) {
+          TimeRange currEventTimeRange = event.getWhen(); 
+        
+          handleOverlapTimeRange(currEventTimeRange, busyTimes);
+
+          // If any one of the mandatory attendees is in this event's guest list,
+          // this event's time range needs to be considered busy. 
+          // Breaks out of the for-each loop because there is no need to look
+          // at the rest of the event's guest list. 
+          break; 
+        }
+      }
+    }); 
+    return busyTimes; 
+  }
+
+  /**
+   * Compares a given event's time slot with all the existing busy time slots.
+   * Merges the current event time slot with a busy time slot if they overlap.
+   * Adds the current event's time slot to the list of busy slots in the end. 
+   * @param current event's time range; a collection of busy time slots
+   */ 
+  private void handleOverlapTimeRange(TimeRange currEventTimeRange, Collection<TimeRange> busyTimes) {
+    boolean overlapped = false; 
+
+    for (TimeRange busyTimeRange : busyTimes) {
+      if (currEventTimeRange.overlaps(busyTimeRange)) {
+        TimeRange combinedTimeRange = getCombinedTimeRange(currEventTimeRange, busyTimeRange);
+        busyTimes.remove(busyTimeRange);
+        busyTimes.add(combinedTimeRange);
```

Comment:
Here you are making changes to busyTimes as you are iterating over busyTimes. Instead, consider updating handleOverlapTimeRange so it returns a new collection of busy time ranges, without modifying the given busyTimes. Then perhaps we can rename handleOverlapTimeRange to,

Collection<TimeRange> getUpdatedBusyTimeRanges